Adds generic block record readers

rubanm opened the pull request:
This resurrects @ianoc's branch that adds generic block record reader, which uses the provided `BinaryConverter` to deserialize data.

Also moved some common code to a new `LzoBlockRecordReader` class.

Will create scalding PR to add `LzoGenericScheme` once this is merged.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/LzoBinaryBlockRecordReader.java:
```diff
@@ -28,81 +20,12 @@
-   */
-
-  private final BinaryBlockReader<M> reader_;
+    extends LzoBlockRecordReader<M, W, BinaryBlockReader<M>> {
```
Is this just a renaming of LzoBinaryBlockRecordReader to LzoBlockRecordReader?

rubanm replied:
Yes, most of the common code has been moved to this class with `nextKeyValue()` left to the implementing classes.

rangadi replied:
I see, who else extends LzoBlockRecordReader?

ianoc replied:
LzoGenericBlockRecordReader extends it also

rubanm replied:
The new LzoGenericBlockRecordReader does.

johnynek commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/GenericBlockReader.java:
```diff
@@ -0,0 +1,156 @@
+      return -1;
+    }
+
+    return b | (in_.read() << 8) | (in_.read() << 16) | (in_.read() << 24);
```
why are we only checking for -1 on the first call to read?

Is this even needed? The -1 | (-1 << 8) | (-1 << 16) | (-1 << 24) == -1. So, everything is fine if you don't do anything, assuming negative values are not allowed.

rubanm replied:
The only reason I can think of is we don't want the subsequent read() calls with how the block markers are set up.

If the first byte is -1, return. Else sum up the next 3 bytes to get the block size. -1 perhaps denotes an empty block and not eof?

ianoc replied:
_in is an input stream, so i believe the contract is that it must mean EOF. @johnynek i figure some presumption that you avoid the extra read calls once you already know its EOF. But given EOF is not on the hot path, and calling read 3 times can't be very expensive its probably a premature optimization maybe? given any call to read can produce -1, might aswell just do the OR and be done with it.

johnynek commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/GenericBlockReader.java:
```diff
@@ -0,0 +1,156 @@
+      return null;
+    }
+
+    int blockSize = readInt();
```
shame this does not use a varint encoding. After compression it won't matter, but before compression we are probably loosing a few percent for records in the 100 byte scale.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/LzoGenericBlockRecordReader.java:
```diff
@@ -0,0 +1,106 @@
+/**
+ * Generic lzo block record reader that uses the provided BinaryConverter to deserialize data.
+ */
+public class LzoGenericBlockRecordReader<M>
```
Overall this replicates a lot of logic from LzoBinaryBlockReader. lets see we can merge these two.

rubanm commented:
Based on offline conversation with @rangadi and @ianoc, dropping the new generic class in favor of moving this logic to the existing `LzoBinaryBlockRecordReader`. This class now 1) skips null records, 2) deserializes the read ByteString instead of deferring to the underlying `BinaryBlockReader`.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/BinaryConverterProvider.java:
```diff
@@ -0,0 +1,12 @@
+ * A simple interface to serialize and deserialize objects
+ */
+public interface BinaryConverterProvider<M> {
+  BinaryConverter<M> getProvider(Configuration conf);
```
should be getConverter()?

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/LzoBinaryBlockRecordReader.java:
```diff
@@ -129,38 +137,36 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
 
-      if (updatePosition) {
+      if (decoded != null) {
+        if (updatePosition) {
```
I think updatePosition does no depend on decoded, should be outside.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/LzoBinaryBlockRecordReader.java:
```diff
@@ -129,38 +137,36 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
         // As a consequence of this, next split reader skips at least one byte
         // (i.e. skips either partial or full record at the beginning).
       }
+      ByteString bs = reader_.readNextProtoByteString();
```
we are not taking any advantage of ByteString here (as opposed to using byte[]). the intention might have been to avoid an array copy. lets see if that can still happen.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryBlockReader.java:
```diff
@@ -30,19 +29,26 @@
   private boolean skipEmptyRecords = true; // skip any records of length zero
 
-  protected BinaryBlockReader(InputStream in, BinaryConverter<M> protoConverter) {
+  private byte[] lastBlock = null;
```
assign it to byte[1] here itself
It could be renamed. 'lastBlock' kind of implies it retains state perhaps used next block read. But it doesn't may be 'blockBuffer' is better. It used just to read the bytes in.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryBlockReader.java:
```diff
@@ -149,7 +172,7 @@ public long skipToNextSyncPoint() throws IOException {
     }
 
-    SerializedBlock block = SerializedBlock.parseFrom(byteArray);
+    SerializedBlock block = SerializedBlock.parseFrom(byteArray, blockSize);
```
Actually if we use ```in_``` SerializedBlock.parseFrom(in_), then we could remove both byteArray and lastBlock.

rubanm replied:
We'll lose the ability to cache the allocated byte array in the reader here. This optimization was added in this change so we can possibly drop it @ianoc ?

ianoc replied:
I think @rangadi is suggesting we should be able to parse right out of the input stream  -- that would avoid the allocation entirely

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/SerializedBlock.java:
```diff
@@ -77,6 +79,19 @@ public static SerializedBlock parseFrom(byte[] messageBuffer)
             .build());
   }
 
+  public static SerializedBlock parseFrom(byte[] messageBuffer, int limit)
```
Lets add parseFrom(InputStream in) instead.

rangadi commented:
Thanks Ruban, the patch is much smaller now.

You need to update ThriftConverter and ProtobufConverter for new contract: null implies skip, any decoding errors should result in a exception.

You also need to update B64 reader to comply to new contract.

rubanm commented:
@rangadi Thanks for the review. Addressed most of your comments.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/LzoBinaryB64LineRecordReader.java:
```diff
@@ -86,6 +87,7 @@ public void initialize(InputSplit genericSplit, TaskAttemptContext context)
     recordErrorsCounter = HadoopUtils.getCounter(context, group, "Errors");
     emptyLinesCounter = HadoopUtils.getCounter(context, group, "Empty Lines");
     truncatedLinesCounter = HadoopUtils.getCounter(context, group, "Truncated Lines");
+    recordsSkippedCounter = HadoopUtils.getCounter(context, group, "Null Records Skipped");
```
"Records Skipped" (this could be for any reason, e.g. dropped by a push down filter). Or you could say "Records Skipped by Converter"

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryBlockReader.java:
```diff
@@ -141,15 +154,13 @@ public long skipToNextSyncPoint() throws IOException {
     }
 
-    SerializedBlock block = SerializedBlock.parseFrom(byteArray);
+    SerializedBlock block = SerializedBlock.parseFrom(new BoundedInputStream(in_, blockSize));
```
BoundedInutStream() is not strictly required, but good.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryConverter.java:
```diff
@@ -8,7 +8,7 @@
 
   /** Returns deserialized object. A return of null normally implies an error. */ 
-  M fromBytes(byte[] messageBuffer);
+  M fromBytes(byte[] messageBuffer) throws Exception;
```
java doc needs an update.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryWritable.java:
```diff
@@ -196,7 +196,7 @@ public String toString() {
     try {
       msgObj = get();
-    } catch (IllegalStateException e) {
+    } catch (Exception e) {
```
what changed?

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/MultiInputFormat.java:
```diff
@@ -120,6 +126,26 @@ public static void setClassConf(Class<?> clazz, Configuration conf) {
+
+      BinaryConverter<?> converter = converterProvider.getConverter(conf);
+      switch (fileFormat) {
+        case LZO_BLOCK:
```
Can you add a TODO : all the three instances of ```switch (fileFormat)``` can be combined into one. We just need to set the right writable and converter for each of the cases (for thrift, protobuf, generic, 'binary' etc).

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryBlockReader.java:
```diff
@@ -79,10 +83,8 @@ public boolean readNext(BinaryWritable<M> writable) throws IOException {
-   */
-  public byte[] readNextProtoBytes() throws IOException {
+
+  public ByteString readNextProtoByteString() throws IOException {
```
It does not look like this change is necessary. Leaving readNextProtoBytes() as it was is better until we gain from working on ByteString directly.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/ProtobufConverter.java:
```diff
@@ -62,10 +63,11 @@ public M fromBytes(byte[] messageBuffer, int offset, int len) {
                                 .build();
     } catch (InvalidProtocolBufferException e) {
```
we can remove this exception logging since the exception is percolated up.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/SerializedBlock.java:
```diff
@@ -69,6 +73,16 @@ public static SerializedBlock newInstance(String protoClassName, List<ByteString
 
+  public static SerializedBlock parseFrom(InputStream in)
+                                          throws InvalidProtocolBufferException, IOException {
+    // note this reads the entire input stream so it should
```
It would read the entire inputstream only in case of corrupt data, right?

rubanm replied:
This throws if I send the unbounded stream actually. This seems to suggest it always reads the entire stream:
https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/Message.Builder#mergeFrom%28java.io.InputStream%29

```
Caused by: com.google.protobuf.InvalidProtocolBufferException: Protocol message end-group tag did not match expected tag.
	at com.google.protobuf.InvalidProtocolBufferException.invalidEndTag(InvalidProtocolBufferException.java:94)
	at com.google.protobuf.CodedInputStream.checkLastTagWas(CodedInputStream.java:124)
	at com.google.protobuf.AbstractMessageLite$Builder.mergeFrom(AbstractMessageLite.java:211)
	at com.google.protobuf.AbstractMessage$Builder.mergeFrom(AbstractMessage.java:904)
	at com.twitter.elephantbird.mapreduce.io.SerializedBlock.parseFrom(SerializedBlock.java:80)
```

rangadi replied:
I see. sounds good.

rangadi commented:
mostly looks good. added a few more comments. please look into the travis ci failures.

rubanm commented:
Addressed further comments and some build fixes.

rangadi commented:
:+1:  LGTM. Thanks Ruban.

ianoc commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/LzoGenericB64LineRecordReader.java:
```diff
@@ -0,0 +1,19 @@
+
+  public LzoGenericB64LineRecordReader(TypeRef<M> typeRef, BinaryConverter<M> converter) {
+    super(typeRef, new GenericWritable<M>(converter), converter);
+    LOG.info("record type is " + typeRef.getRawClass());
```
Don't think this adds anything anymode, the typeRef will just be the convertors creator. I'd just kill the log line

ianoc commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/LzoGenericBlockRecordReader.java:
```diff
@@ -0,0 +1,32 @@
+    super(typeRef,
+      new BinaryBlockReader(null, binaryConverter),
+      new GenericWritable<M>(binaryConverter));
+    LOG.info("LzoGenericBlockRecordReader, type args are " + typeRef.getRawClass());
```
kill

ianoc commented:
Before we merge have you run a sanity job with some data at scale just to make sure it works on hadoop cluster?

laurentgo commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/ProtobufBlockReader.java:
```diff
@@ -38,7 +38,7 @@ public ProtobufBlockReader(InputStream in, TypeRef<M> typeRef) {
   // for backward compatibility :
 
-  public boolean readProtobuf(ProtobufWritable<M> message) throws IOException {
+  public boolean readProtobuf(ProtobufWritable<M> message) throws Exception {
```
Exception is super generic and is usually an anti-pattern. Any reason for this change?

rubanm replied:
TException does not extend IOException so there wasn't a more specific parent class to set here for both the thrift and profobuf codepaths. A new Exception class could be created to wrap the different ones I guess. Will fix.

laurentgo replied:
sounds safest. Thanks

laurentgo commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/GenericWritable.java:
```diff
@@ -0,0 +1,19 @@
+
+  @Override
+  protected BinaryConverter<M> getConverterFor(Class<M> clazz) {
+    throw new RuntimeException("Unsupported");
```
UnsupportedOperationException is probably a better fit

laurentgo commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryWritable.java:
```diff
@@ -84,7 +84,11 @@ public M get() {
-      return converter.fromBytes(messageBytes);
+      try {
+        return converter.fromBytes(messageBytes);
+      } catch (Exception e) {
```
catching Exception looks an antipattern. It would catch OutOfMemoryException and any other RuntimeException and the job might continue but in such a bad state that it would probably be a waste of resources

rangadi replied:
other alternative is to require throwing a new exception like 'DecodeException' in the converter interface.

laurentgo commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryWritable.java:
```diff
@@ -196,7 +200,7 @@ public String toString() {
     try {
       msgObj = get();
-    } catch (IllegalStateException e) {
+    } catch (Exception e) {
```
Same comment regarding catching Exception

laurentgo commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryBlockReader.java:
```diff
@@ -43,6 +44,10 @@ protected BinaryBlockReader(InputStream in,
     this.skipEmptyRecords = skipEmptyRecords;
   }
 
+  public BinaryConverter<M> getProtoConverter() {
```
if this class is generic and work with thrift, references to protobuf should be changed in argument and method names

rangadi replied:
Yeah, it has always been that way. proto could be interpreted as generic protocol of how a message is serialized (protobuf, thrift, etc)

rubanm replied:
Leaving this unchanged. Let me know if you feel strongly about the rename though.

laurentgo commented on core/src/main/java/com/twitter/elephantbird/mapreduce/input/LzoBinaryBlockRecordReader.java:
```diff
@@ -129,22 +135,20 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
-        // Re-throw IOExceptions that are not due to protobuf decode errors
-        throw e;
+        decoded = deserializer_.fromBytes(byteArray);
       } catch (Throwable e) {
```
not new code, but it means it would catch any RuntimeException like OOM

rangadi replied:
We have to assume we have corrupt bytes that we are trying to deserialize. Thrift can throw any exception.. we try to avoid OOM with our own wrapper, but there is no guarantee. Given the requirement that we should be able to ignore the corrupt records, what is the solution?

rubanm replied:
Seems like we should change `Throwable` to `Exception` at least? Doesn't fix the case Laurent pointed out but slightly better.

rubanm commented:
@laurentgo addressed most of your comments, thanks.

@ianoc ran a day's worth of client event logs and a couple of other sources including protobuf ones as a sanity check.

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/ThriftConverter.java:
```diff
@@ -46,18 +48,15 @@ public ThriftConverter(TypeRef<M> typeRef) {
-      // other runtime exceptions (e.g. IndexOutOfBoundsException).
-      logWarning("failed to deserialize", e);
-      return null;
+    } catch (TException e) {
```
This is not enough comment from deleted code :

// normally a TException. but some corrupt records can cause
// other runtime exceptions (e.g. IndexOutOfBoundsException).

rangadi commented on core/src/main/java/com/twitter/elephantbird/mapreduce/io/BinaryConverterDecodeException.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Thrown by BinaryConverter if it fails to deserialize bytes.
+ */
+public class BinaryConverterDecodeException extends IOException {
```
This is added here, but not used on the read side (Lzo*RecordReader). Not sure if this gives us much. But I am ok if you want to keep it.

rubanm replied:
The record readers already declare IOException which is why it works without adding there I think. I can make it explicit though.

laurentgo commented:
@rubanm lgtm
